### PR TITLE
Permutation interface

### PR DESCRIPTION
**Pull request by dgregoire**
Kensler permutation & permutation interface

### CODE REVIEW

File: internal/permutations/kensler.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+/*
+
+Andrew Kensler, a researcher at Pixar, introduced an interesting technique for generating the permutation of an array
+in his 2013 paper, Correlated Multi-Jittered Sampling.
+
+reference:             https://graphics.pixar.com/library/MultiJitteredSampling/paper.pdf
+further comments from: https://afnan.io/posts/2019-04-05-explaining-the-hashed-permutation/
+
+*/
```
Comment (juanli16) [inline internal/permutations/kensler.go:18]: Do you want this to be a package comment instead maybe?

File: internal/permutations/kensler.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	return kensler{l: uint32(l), p: uint32(i.Int64())}, nil
+}
+
+// Shuffle using the kensler algorithm
+// with n the number to permute/the index of the permutation vector.
+// This is not totally appropriate for our application
+// since this only works on uint32 size.
+//
+// As long as the number of items being matches is not >4b
```
Comment (juanli16) [inline internal/permutations/kensler.go:49]:
```suggestion
// As long as the number of items being matched is not >4b
```

File: internal/permutations/kensler.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	var max = big.NewInt(l - 1)
+	i, err := rand.Int(rand.Reader, max)
+	if err != nil {
+		return kensler{}, err
+	}
+
+	return kensler{l: uint32(l), p: uint32(i.Int64())}, nil
+}
+
+// Shuffle using the kensler algorithm
```
Comment (juanli16) [inline internal/permutations/kensler.go:44]:
```suggestion
// Shuffle uses the Kensler algorithm
```

File: internal/permutations/permutations_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		seq++
+
+		if seq == max {
+			/*
+				for _, pos := range enc.permutations[enc.sent:] {
+					if _, err = enc.w.Write(enc.b[pos][:]); err != nil {
+						return
+					}
+				}
+			*/
```
Comment (juanli16) [inline internal/permutations/permutations_test.go:43]: still needed?

File: pkg/dhpsi/dhpsi.go [hunk 5]
```diff
@@ -49,9 +51,11 @@ func NewDeriveMultiplyShuffler(w io.Writer, n int64, gr Ristretto) (*DeriveMulti
 	if err := binary.Write(w, binary.BigEndian, &n); err != nil {
 		return nil, err
 	}
-	// and create the encoder
-	return &DeriveMultiplyShuffler{w: w, max: n, gr: gr, permutations: initP(n), b: make([][EncodedLen]byte, n)}, nil
-
+	// create the permutations
+	p, _ := permutations.NewKensler(n)
+	// and create the buffer map & encoder
+	b := make(map[int64][EncodedLen]byte, int(float64(n)*0.75))
```
Comment (juanli16) [inline pkg/dhpsi/dhpsi.go:57]: Is there a reason for using 3/4 of the size?

**Review by juanli16 — COMMENTED** [review]
Just a few questions.

File: internal/permutations/kensler.go (hunk 1 above)
Comment (dgregoire) [reply internal/permutations/kensler.go:18]: sure, wanna update it?

File: internal/permutations/permutations_test.go (hunk 4 above)
Comment (dgregoire) [reply internal/permutations/permutations_test.go:43]: I guess not, its a template for the code just under

File: pkg/dhpsi/dhpsi.go (hunk 5 above)
Comment (dgregoire) [reply pkg/dhpsi/dhpsi.go:57]: during testing I noticed it tended to fill 75% of the map but its a very thin link. Maybe half or full would be better

Comment (juanli16) [reply pkg/dhpsi/dhpsi.go:57]: I see, but do you need this optimization for the Parallel version as well?
also, maybe just add an one line comment to state that.

File: internal/permutations/kensler.go (hunk 1 above)
Comment (juanli16) [reply internal/permutations/kensler.go:18]: yeh, it's a nice description of what this package is and where it comes from.

File: pkg/dhpsi/dhpsi.go (hunk 5 above)
Comment (dgregoire) [reply pkg/dhpsi/dhpsi.go:57]: the parallel version doesn't implement caching at this time

**Review by juanli16 — APPROVED** [review]
Good work!

**Pull request merged**